src/rgw/rgw_op.cc

```diff
@@ -4568,6 +4568,23 @@ void RGWPutACLs::execute()
     return;
   }
 
+  const RGWAccessControlList& req_acl = policy->get_acl();
+  const multimap<string, ACLGrant>& req_grant_map = req_acl.get_grant_map();
+#define ACL_GRANTS_MAX_NUM      100
+  int max_num = s->cct->_conf->rgw_acl_grants_max_num;
+  if (max_num < 0) {
+    max_num = ACL_GRANTS_MAX_NUM;
+  }
+
+  int grants_num = req_grant_map.size();
+  if (grants_num > max_num) {
+    ldout(s->cct, 0) << __func__ << " An ACL can have up to 100 grants, request ACL grants num: "
+                     << grants_num << dendl;
+    op_ret = -ERR_MALFORMED_ACL_ERROR;
+    s->err.message = "The XML you provided was not well-formed or did not validate against our published schema";
```

Comment:
similar to feedback on https://github.com/ceph/ceph/pull/16310:
* raise debug level to 4
* can we give a better error message here? should say that it's rejected because it exceeds the maximum number of acls